### PR TITLE
Sprinting, Stamina and Temporary Modifiers

### PR DESCRIPTION
**Pull request by NanakoAC**
First draft of a major new feature, in need of feedback and playtesting.
Not quite perfected yet

IMPORTANT
In order to work as intended, this also requires some changes to the walk speed and human move delay in config files. I'm currently using:
WALK_SPEED  4
HUMAN_DELAY 2.5

These are both found in game_options.txt

In addition, Tickcompensation has been disabled. As per a discussion with skull, he believed it was already disabled, but it actually was active. The fact that he believed it wasn't is evidence of its worthlessness as a feature, and i think movement is smoother without it.

As per the description, Tickcomp 'attempts' to smooth out movement. I contend that it utterly fails in this attempt. All it does is add a small, flat value to every movement step, and it serves no purpose but to slow down movement without any smoothing. We may be well served trying to improve it to use some lag-dependant variable like tick usage in future. But for now it breaks the sprinting system by damping high speeds too much.

**Main changes:**

Walk speed increased to match what we currently know as 'run', and made the default movement mode.

Run mode has been changed into a limited high speed sprinting mode. Sprinting works slightly differently for each species, and is intended to be balanced asymmetrically:

Humans have the largest stamina pool, and can run for longest. We evolved as endurance hunters
Unathi have the smallest stamina pool, but the fastest sprint speed. Their sprint is a shortranged burst evolved for surprising prey
Tajarans have very poor sprinting., sprinting is a waste of energy and they evolved to maintain a steady pace and conserve energy in their wintery homeworld.
Skrell are intellectuals and have evolved a reduced need for physical prowess, theyre slightly better than tajarans but still subpar
Vaurca sprinting for now is just slightly faster than humans, may need some creative input.
IPCs, being created by and in the image of humans, have almost the same sprinting speed. As machines they have no stamina, but they will overheat and consume power rapidly while sprinting.
Diona remain poderous, they have the slowest sprint speed of all, and their sprinting consumes stored light energy and radiation,  rather than stamina

Whenever anyone reaches their limit, they will automatically drop to walk speed, and usually collapse in pain or some approximation of it. IPCs are punished far more by this, they take a full minute to initialise emergency backup power

For most organic species, the limit isn't the end o the stamina pool. When stamina runs out they'll start taking suffocation damage up to a point and then collapse.

In addition, when in sprint mode, the run button flashes a colour, gradiating between green/yellow/orange/red to indicate remaining stamina. Sprinting while its red will hurt. 
The best way i could find to do this, since overlay images can't really be updated, is repeatedly removing and re-adding the overlay at the new colour while sprinting. This is quite a performance intensive process, but i think its acceptable given the limited nature of sprinting

Also added a walkspeed limiting, allowing the user to precisely select or input a limit for their walking speed. This allows fast people to slow down, or for people to roleplay slowness, limping and injury. Check IC tab or altclick run/walk button

This PR also adds the first iteration of my new Modifiers code, a system designed to add temporary, reversible changes to anything. Its used so far to add three effects for chemicals which would have been hard to do robustly without such a system.

The effects added for it, give some move/sprintspeed and stamina interactions to several chemicals - inaprovaline, synaptizine, hyperzine, alcohol, and caffienated drinks of all varieties. The effects within a category (EG, stimulants from caffiene and hyperzine) don't stack with each other, only the strongest application of that effect is applied

### CODE REVIEW

**Comment by skull132** [comment]
Does the presence of tick compensation actually hinder the development of this system, or do you simply think it to be worthless?

Because, it's a config option, so using it is up to the user. If it makes certain mechanics not as pronounced, then that the user's call.

**Comment by NanakoAC** [comment]
RE: Tickcomp, yes it does. Because it adds a flat value to the move delay, it becomes worse the lower your delay goes, and largely cripples sprinting.

RE: 'the user'. Can you clarify this? We are a fork of baystation, they added that for server admins to use in configuring their own server. As far as i'm aware, even if it is opensource, aurora isn't intended for general distribution and we make no effort to support potential other servers which might take and host our code. 

Regardless, if that were the intention, i say that tickcomp in its current form is unfit for purpose. The only purpose it serves is adding a tiny bit of overhead by requiring an if check during every movement of every client mob. Which i'd say is enough of a reason to entirely remove/comment out the code, rather than just disabling it in config. 

And to be clear, i'm not opposed to the concept of movement smoothing, only to this ham-fisted attempt at it. I'd definitely be in support of rewriting it at some point

I will contact the lore team for feedback on individual species before this PR is finalised
And yes tajarans still have a faster walking speed

RE: The removal of the existing walk: That is one of the things i wanted feedback about.
Regarding the old walk speed, i only know of one case, ever, that it was actually used as walking - Kaed playing Ickthar Slithiss used to always move in walk mode, to emphasis the limping nature of his old-man character. However the purpose of that was just to differentiate from other characters, and i believe under this new system such a character would simply not use sprint.

The only potential problem i see is slipping - with walk being the new default mode, people will generally never slip on water anymore. I was thinking of adding slipping on wet floors - perhaps with some percentage chance, to walking mobs as well, but that would strip people of a way to reliably avoid it. A slow walk could be a good compromise for that

But i have an alternate idea i'd like your thoughts on. Adding a verb/option, in the IC tab, to set your walk speed, as a percentage of your usual. People could granularly set it to any specific amount they want from 0-100% and thus walk at the specific slowness they want. Then i could have floor slipping being dependant on the actual speed you're moving

File: code/_onclick/hud/screen_objects.dm [hunk 1]
```diff
@@ -589,3 +588,23 @@
...
+					icon_state = "running"
+					if (user.max_stamina != -1)//If max stamina is -1, this species doesnt use stamina. no overlay for them
+						var/image/holder = image('icons/mob/screen1.dmi', src, "run_overlay")
+						var/staminaportion = user.stamina / user.max_stamina
+						holder.color = percentage_to_colour(staminaportion)
+						holder.blend_mode = BLEND_MULTIPLY
+						overlays += holder
+
+				if("walk")
+					overlays.Cut()
```
Comment (skull132) [inline code/_onclick/hud/screen_objects.dm]: This overlays.Cut() is unnecessary, due to the one on line 597.

**Comment by Fire-and-Glory** [comment]
'Unathi have the smallest stamina pool, but the fastest sprint speed.'

'Tajarans have very poor sprinting.'

Rip my current lizard Paramedic and plans for Tajaran Paramedic.

Jests aside, I do suggest a slow-walk, it doesn't sound like something cripplingly hard to implement, and for things like wet floors it could be the "safe" option while the 'fast-walk' has a chance to make movement problems and sprint is very high chance to fall on your ass, along with other potential uses.

**Comment by NanakoAC** [comment]
@Fire-and-Glory Slow walk function is in, and better than expected,

Tajarans still retain their species movespeed bonus, and anyone will be able to get a bit of extra speed if needed via chem abuse

File: code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Dispenser.dm [hunk 2]
```diff
@@ -65,6 +65,8 @@
 	var/adj_temp = 0
 	var/targ_temp = 310
 	var/halluci = 0
+	var/datum/modifier/caffiene_mod
```
Comment (Bedshaped) [inline code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Dispenser.dm]: fyi it's "caffeine"

File: code/game/turfs/turf.dm [hunk 3]
```diff
@@ -145,14 +145,15 @@
 		if(istype(MOB) && !MOB.lying && footstep_sound)
 			if(istype(MOB.shoes, /obj/item/clothing/shoes) && !MOB.shoes:silent)
 				if(MOB.m_intent == "run")
+					playsound(MOB, footstep_sound, 70, 1)
```
Comment (Bedshaped) [inline code/game/turfs/turf.dm]: too loud perhaps?

File: code/__HELPERS/unsorted.dm [hunk 4]
```diff
@@ -1370,3 +1370,21 @@ var/mob/dview/dview_mob = new
 // call to generate a stack trace and print to runtime logs
 /proc/crash_with(msg)
 	CRASH(msg)
+
+/atom/proc/find_up_hierarchy(var/atom/target)
```
Comment (skull132) [inline code/__HELPERS/unsorted.dm]: This proc doesn't appear to be recursive? A recursive proc would reference itself. Right now, it appears to only run one actual cycle 10 times?

File: code/controllers/Processes/modifier_process.dm [hunk 5]
```diff
@@ -0,0 +1,28 @@
+#define MODIFIER_PROCESS_INTERVAL	10
```
Comment (skull132) [inline code/controllers/Processes/modifier_process.dm]: Unless you need this elsewhere, just don't use a `#define` for it. No other processes have a define directive for their interval, they just use number literals.

File: code/controllers/Processes/modifier_process.dm [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+	..()
+	if(!processing_modifiers)
+		processing_modifiers = list()
+
+/datum/controller/process/modifier/doWork()
+	for(last_object in processing_modifiers)
+		var/datum/modifier/O = last_object
+		if(isnull(O.gcDestroyed))
+			try
+				O:process()
```
Comment (skull132) [inline code/controllers/Processes/modifier_process.dm]: This grab shouldn't be necessary, if you've already got `/datum/modifier/proc/process()` defined. See about removing it, and replacing it with a simple `.` operator.

File: code/game/modifiers/modifiers.dm [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+#define MODIFIER_OVERRIDE_REFRESH 	3
+//This mode will overwrite the variables of the old one with our new values
+//It will also force it to remove and reapply its effects
+//This is useful for applying a lingering modifier, by refreshing its duration
+
+#define MODIFIER_OVERRIDE_STRENGTHEN	4
+//Almost identical to refresh, but it will only apply if the new modifer has a higher strength value
+//If the existing modifier's strength is higher than the new one, the new is discarded
+
+#define MODIFIER_OVERRIDE_CUSTOM	4
```
Comment (skull132) [inline code/game/modifiers/modifiers.dm]: Is this supposed to be 5?

File: code/game/modifiers/modifiers.dm [hunk 8]
```diff
@@ -0,0 +1,446 @@
...
+	return 1
+
+
+/datum/modifier/proc/handle_registration(var/override = 0)
+	var/datum/modifier/existing = null
+	for (var/datum/modifier/D in target.modifiers)
+		if (D.type == type)
+			existing = D
+	if (!existing)
+		processing_modifiers |= src
```
Comment (skull132) [inline code/game/modifiers/modifiers.dm]: `|=` operator is slow. It goes through the entire list, looking for `src` in there, and if it doesn't find it, it adds it. Think you can write it to be safe enough to use the cheaper `+=` operator?

File: code/modules/mob/living/carbon/human/species/species.dm [hunk 9]
```diff
@@ -310,3 +316,35 @@
 // Called in life() when the mob has no client.
 /datum/species/proc/handle_npc(var/mob/living/carbon/human/H)
 	return
+
+/datum/species/proc/handle_sprint_cost(var/mob/living/carbon/human/H, var/cost)
+	cost *= H.sprint_cost_factor
+	if (H.stamina == -1)
+		log_debug("Error: Species with special sprint mechanics has not overridden cost function.")
```
Comment (skull132) [inline code/modules/mob/living/carbon/human/species/species.dm]: Just pointing this one out. If you need it in live code, then keep it.

File: code/__HELPERS/unsorted.dm (hunk 4 above)
Comment (NanakoAC) [inline code/__HELPERS/unsorted.dm]: Woops, forgot the line where it sets itself to its location

File: code/controllers/Processes/modifier_process.dm (hunk 5 above)
Comment (NanakoAC) [inline code/controllers/Processes/modifier_process.dm]: done

File: code/controllers/Processes/modifier_process.dm (hunk 6 above)
Comment (NanakoAC) [inline code/controllers/Processes/modifier_process.dm]: I figured this was done to prevent one bad modifier from crashing all of them and breaking the system, a robustness measure.

not necessary im sure, done

File: code/game/modifiers/modifiers.dm (hunk 7 above)
Comment (NanakoAC) [inline code/game/modifiers/modifiers.dm]: whoops!

File: code/game/modifiers/modifiers.dm (hunk 8 above)
Comment (NanakoAC) [inline code/game/modifiers/modifiers.dm]: im sure it already is, yea that modifier was unnnecessary

File: code/modules/mob/living/carbon/human/species/species.dm (hunk 9 above)
Comment (NanakoAC) [inline code/modules/mob/living/carbon/human/species/species.dm]: yes it's good to keep, thats why i switched it to log debug. i use world << while testing

**Comment by skull132** [comment]
Oh.

And please adhere to naming convention. "modifier_process.dm" should be "modifier.dm" or "modifiers.dm". All process files are named after the process name, "_process" is already implied by the file's presence in the Processes folder.

File: code/datums/datumvars.dm [hunk 10]
```diff
@@ -952,3 +952,6 @@ client
 		src.debug_variables(DAT)
 
 	return
+
+/datum/proc/process_tick()
```
Comment (skull132) [inline code/datums/datumvars.dm]: Why is this necessary? It doesn't appear to be referenced anywhere else.

File: code/modules/mob/living/living_powers.dm [hunk 11]
```diff
@@ -32,10 +32,34 @@
...
+
+	var/response = input(src, "Your current walking speed is [speed] tiles per second. This menu allows you to limit it to a lower value, by applying a multiplier to that. Please choose a value, select custom to enter your own, or No limit to set your walk speed to the maximum. This menu will not make you move any faster than usual, it is only for allowing you to move at a slower pace than normal, for roleplay purposes. Values set here will not affect your sprinting speed", "Limit Walking speed") as null|anything in options
+	if (response == "No limit")
+		src << "Movement speed has now been set to normal, limits removed."
+		src.min_walk_delay = 0
+		return
+	else if (response == "Custom")
+		response = input(src, "Please enter the exact speed you want to walk at, in tiles per second. This value must be less than [speed] and greater than zero", "Limit Walking speed") as num
+		newspeed = response
+	else
```
Comment (skull132) [inline code/modules/mob/living/living_powers.dm]: This will create issues.

`null|anything` means that the selection can be exited out of with "Cancel", which gives response a value of `null`. Attempting to convert that into a number and then using it in later operations will bring up issues.

So you should have a catch for it:
```
else if (isnull(response))
    return
```

Alternatively, do this:
```
else if (isnum(response))
    newspeed = text2num(response)
else
    return
```

May want to test, but I _think_ DM recognizes a string that only contains digits as a number. I think.

File: code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Dispenser.dm (hunk 2 above)
Comment (skull132) [inline code/modules/reagents/Chemistry-Reagents/Chemistry-Reagents-Dispenser.dm]: Since changes are required again. Please also fix this grammar error.

File: code/modules/reagents/Chemistry-Reagents.dm [hunk 12]
```diff
@@ -61,6 +65,11 @@
 	remove_self(removed)
 	return
 
+
+//Initial effect is called once when the reagent first starts affecting a mob.
+/datum/reagent/proc/initial_effect(var/mob/living/carbon/M, var/alien)
```
Comment (skull132) [inline code/modules/reagents/Chemistry-Reagents.dm]: Isn't referenced anywhere else, but in changes to this file.

Null change, so please remove.

File: code/_onclick/hud/screen_objects.dm (hunk 1 above)
Comment (skull132) [inline code/_onclick/hud/screen_objects.dm]: Please also address this. It's a minor thing, but, it's an obvious thing.

File: code/modules/mob/living/living_powers.dm (hunk 11 above)
Comment (NanakoAC) [inline code/modules/mob/living/living_powers.dm]: fixing

File: code/modules/reagents/Chemistry-Reagents.dm (hunk 12 above)
Comment (NanakoAC) [inline code/modules/reagents/Chemistry-Reagents.dm]: i added this because its quite needed and a significant hole in functionality, it will be useful in future, best to keep it

File: code/modules/mob/living/living_powers.dm [hunk 13]
```diff
@@ -32,10 +32,34 @@
...
 
-/*
-/mob/living/verb/devourverb(var/mob/living/victim)//For situations where species inherent verbs isnt suitable
-	set category = "Abilities"
-	set name = "Devour Creature"
-	set desc = "Attempt to eat a nearby creature, swallowing it whole if small enough, or eating it piece by piece otherwise"
-	attempt_devour(victim, eat_types, mouth_size)
-*/
\ No newline at end of file
+/mob/living/verb/set_walk_speed()
```
Comment (NanakoAC) [inline code/modules/mob/living/living_powers.dm]: i dont see any way you could activate this menu on someone else's behalf, it doesnt seem like it'd affect anything
